File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        rows, cols, d, heap = len(matrix), len(matrix[0]), [], []
        length = rows * cols
        
        for y in matrix:
            for x in y:
                heap.append(x)
                
        heapq.heapify(heap)
      
        for i in range(k-1):
            heapq.heappop(heap)
            
        return heap[0]

File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int n = matrix.size();
        auto f = [&](int x) {
            int cnt = 0;
            for (int i = 0; i < n; i++) {
                int j = upper_bound(matrix[i].begin(), matrix[i].end(), x) - matrix[i].begin();
                cnt += j;
            }
            return cnt >= k;
        };
        int lo = matrix[0][0], hi = matrix[n-1][n-1];
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (f(mid))
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo;
    }
};